feat: dao launcher form wireup

ashmidgley opened the pull request:
[DAO Launcher Governance Step](https://www.notion.so/rational-nomads/DAO-Launcher-Governance-Step-4a4fa4ae8aca4b9e8182013c812ec806)
[DAO Launcher Launch Step](https://www.notion.so/rational-nomads/DAO-Launcher-Launch-Step-d2c5c94568f543fea9a791e7a1fe1d28)
[DAO Launcher Treasury Step](https://www.notion.so/rational-nomads/DAO-Launcher-Treasury-Step-51d9c67d5c24418aa392aba5965b3f7b)

-----

## 1. Pull request checklist
- [x] Notion card has been moved to the Code Review column
- [x] Notion card has a link to this PR
- [x] A reviewer has been assigned to the Notion card

## 2. PR type
```
- Feature
```

## 3. What is the old behaviour?
WrappedToggleInput, WrappedDropdownMenu and DAO launcher governance step not implemented.

## 4. What is the new behaviour?
Addition of governance form and new components to DAO launcher.

colbr commented on src/features/create-dao/CreateDAO.types.tsx:
```diff
@@ -0,0 +1,89 @@
+	},
+];
+
+export const getVotingPeriodItems = (setFieldValue: any): DropdownItem[] =>
```
nit: square brackets for these - makes them a little easier to parse. I'll make a task for adding this to our lint rules ☺️

colbr commented on src/features/create-dao/steps/DAOSummaryField.tsx:
```diff
@@ -0,0 +1,21 @@
+	label,
+	value,
+}) => (
+	<div className={classNames(styles.Field, className)}>
+		<p className={styles.FieldLabel}>{label}</p>
+		<p className={styles.FieldValue}>{value}</p>
+	</div>
```
Use [TextDetail](https://zui-storybook.netlify.app/?path=%2Fstory%2Ftypography-text-detail--default) here.

colbr commented on src/features/create-dao/steps/GovernanceForm.test.tsx:
```diff
@@ -0,0 +1,220 @@
+
+// We need to mock ResizeObserver to test the dropdown menu's.
+// See https://github.com/radix-ui/primitives/issues/420#issuecomment-1125837782.
+(global as any).ResizeObserver = class ResizeObserver {
+	constructor(cb: any) {
+		(this as any).cb = cb;
+	}
+
+	observe() {
+		(this as any).cb([{ borderBoxSize: { inlineSize: 0, blockSize: 0 } }]);
+	}
+
+	unobserve() {}
+};
```
Can we move this to the global jest config, so we don't have to add it in each test suite?

colbr commented on src/features/create-dao/steps/GovernanceForm.test.tsx:
```diff
@@ -0,0 +1,220 @@
+		await waitFor(() =>
+			expect(
+				screen.getByText('The DAO token address field is required.'),
+			).not.toBe(null),
```
Use `toBeInTheDocument` here - it will read a bit cleaner

colbr commented on src/features/create-dao/steps/GovernanceForm.module.scss:
```diff
@@ -0,0 +1,26 @@
+	flex-direction: column;
+}
+
+.VotingProcess {
+	width: 50%;
+	padding-right: 1rem;
+}
+
+.VotingPeriod {
+	width: 50%;
+	margin-left: 1rem;
+}
+
+.VotingSystem {
+	margin-bottom: 2rem;
+}
+
+.Footer {
+	margin-top: 5rem;
+}
```
nit: use `gap` here instead of `padding` and `margin`.

colbr commented on src/features/ui/DropdownIconLabel/DropdownIconLabel.tsx:
```diff
@@ -0,0 +1,20 @@
+import { FC } from 'react';
+
+import styles from './DropdownIconLabel.module.scss';
+
+export type DropdownIconLabelProps = {
+    icon: JSX.Element,
+    label: string;
+};
+
+export const DropdownIconLabel: FC<DropdownIconLabelProps> = ({
+	icon,
+    label
+}) => (
+    <div className={styles.Container}>
+        {icon}
+        <div className={styles.Label}>
+            <span>{label}</span>
+        </div>
+    </div>
+);
```
Can we move this functionality to zUI?

domw30 replied:
Oh nice, yes we should move this over as we'll be re-using this plenty of times

ashmidgley replied:
Yep, happy to move! Are you okay for me to address as a seperate piece? Stops this one getting blocked by a zUI update.

colbr commented on src/features/ui/WrappedDropdownMenu/WrappedDropdownMenu.tsx:
```diff
@@ -0,0 +1,44 @@
+	helperText: string;
+};
+
+export const WrappedDropdownMenu: FC<WrappedDropdownMenuProps> = ({
+	className = '',
+	items = [],
+	value,
+	label,
+	placeholder,
+	info,
+	hasError,
+	helperText,
+}) => {
+	const trigger = (
+		<Input type="text" placeholder={placeholder} value={value} onChange={() => {}} error={hasError} helperText={helperText} endEnhancer={<IconChevronDown />} />
+	);
+
+	return (
+		<div className={classNames(styles.Container, className)}>
+			<div className={styles.LabelContainer}>
+				<p className={styles.Label}>{label}</p>
+				<InfoTooltip content={info} />
+			</div>
+			<DropdownMenu trigger={trigger} items={items} />
+		</div>
+	);
+}
```
This are all good for now until zUI has what y'all need. Can we make a generic Wrapper component though? E.g.

```
<Wrapper
    label={label}
    info={info}
>
    <ChildComponent />
</Wrapper>
```

Will make it easier to refactor when the time comes

colbr reviewed (changes requested):
Looking awesome. A few small comments. I'll get @domw30 's opinion as well as he's more up to date on NFTs stuff!

domw30 commented on src/features/create-dao/steps/DAOSummaryField.module.scss:
```diff
@@ -0,0 +1,18 @@
+
+	.FieldLabel {
+		font-size: 0.75rem;
+		color: #{colors.$grey-lighter-2};
```
I could be wrong, but I have a feeling this should throw an error ?? 
```
color: $grey-lighter-2;
```

ashmidgley replied:
Yeah, unsure why this wasn't erroring. Updated!

domw30 commented on src/features/create-dao/steps/TreasuryForm.module.scss:
```diff
@@ -0,0 +1,17 @@
+.InputExtraInfo {
+	text-align: right;
+	font-size: 0.75rem;
+	color: #{colors.$blue};
```
Same here, does this not throw an error in the UI ? Oh I think I see why this would work, you should import these from the `zUI` package, rather than local linking?

ashmidgley replied:
Yep, good shout. Have updated import!

domw30 reviewed (changes requested):
Nice work! Few comments picked out but nothing major 💯

domw30 commented:
I also noticed some UI/styling may be off or missing, but noted from the walk through that this will be addressed in follow up work. I think we will need to re-visit the tooltips and icons used they look a little strange. Tooltip looks like it is visible when opening the container that holds them. I'll make a note to check up on these.

ashmidgley commented on src/features/create-dao/steps/DAOSummary.module.scss:
```diff
@@ -0,0 +1,71 @@
+			height: 12.5rem;
+			width: 12.5rem;
+			margin-right: 2rem;
+			border-radius: 0.5rem !important;
```
Noted! Will add a task to address this in the next lot of work.

ashmidgley commented on src/features/create-dao/steps/DAOSummary.tsx:
```diff
@@ -0,0 +1,120 @@
+import { DAOSummaryField } from './DAOSummaryField';
+
+import styles from './DAOSummary.module.scss';
+import { truncateAddress } from '../../../lib/util/domains/domains';
```
Good to know! Cheers for the heads up.

ashmidgley commented on src/features/create-dao/steps/DAOSummary.tsx:
```diff
@@ -0,0 +1,120 @@
+				value={details.description}
+				className={styles.Text_Description}
+			/>
+			<div className={styles.Text_Confirmation}>
```
Definitely agree. Will add a note for the next piece of work.

ashmidgley commented:
Cheers @colbr and @domw30 for the review! Have addressed all comments. Worth noting:
- If possible I'll look to move the `DropdownIconLabel` into zUI as a separate piece.
- Noted change/improve some of the naming of the styles to avoid using underscores + repeating names + use border variables in `DAOSummary` for our next piece.
- @domw30, @colbr has mentioned previously not to worry about the Tooltip being visible when opening the container that holds them. I think there were plans for it to be addressed by the zUI team?

Let me know if I've missed anything!

domw30 commented on src/features/create-dao/CreateDAO.types.tsx:
```diff
@@ -0,0 +1,92 @@
+};
+
+export const getVotingSystemItems = (
+	setFieldValue: (field: string, value: any, shouldValidate?: boolean) => void,
```
Could value type be `string` here?

domw30 commented on src/features/create-dao/steps/DAOSummary.module.scss:
```diff
@@ -0,0 +1,70 @@
+		
+		.Summary {
+			margin-right: 1.3rem;
+			width: 8rem;
```
I think these styles can be improved to future proof responsive design, but as we've discussed we can circle back on some of the styling

domw30 commented on src/features/create-dao/steps/DAOSummary.module.scss:
```diff
@@ -0,0 +1,70 @@
+}
+
+.Text {
+	font-family: 'Inter';
```
Shouldn't need to set font-family, font-style, font-weight. This look like the defaults.

ashmidgley replied:
Good spotting. Have removed.

domw30 commented on src/features/create-dao/steps/DAOSummary.module.scss:
```diff
@@ -0,0 +1,70 @@
+		height: 4.5rem;
+		border-top: solid 0.063rem #232323;
+		padding: 1.5rem 0;
+		margin-top: 1.25rem;
```
nit: we could reduce to one liners, something like `margin: 1.25rem 0 1rem 0`

domw30 commented on src/features/create-dao/steps/DAOSummary.module.scss:
```diff
@@ -0,0 +1,70 @@
+}
+
+.Footer {
+	margin-top: 1rem;
```
`gap` would be good to use over margins between these sections. Can circle back to this.

domw30 commented on src/features/create-dao/steps/DAOSummaryField.tsx:
```diff
@@ -0,0 +1,25 @@
+		className={classNames(styles.Field, className)}
+		label={label}
+		primaryText={value.toString()}
+		secondaryText=""
```
Ah I've seen this a couple of times in my work now too. I think we need to update `TextStack` in `zUI` to make `secondaryText` optional. Doesn't make sense to have to pass an empty string for the prop here. I'll make a task and follow up

domw30 commented on src/features/create-dao/steps/GovernanceForm.test.tsx:
```diff
@@ -0,0 +1,195 @@
+
+		const user = userEvent.setup();
+
+		await user.click(screen.getByText(/MAJORITY/i));
```
I think you could remove any use of `await` when using `getBy...`.

ashmidgley replied:
The `await` is for the user.click action. Removing breaks the tests:

![Screen Shot 2022-11-25 at 1 07 30 PM](https://user-images.githubusercontent.com/13953984/203876626-1763bc54-a192-484f-b479-58419ed63ffb.png)

domw30 commented on src/features/ui/WrappedToggleInput/WrappedToggleInput.tsx:
```diff
@@ -0,0 +1,42 @@
+	label: string;
+	value: string;
+	info: string;
+	learnMoreHref?: string;
```
I feel `href` would be better than `learnMoreHref`. Just in case design ever changed the text content of the link. Let me know thoughts

ashmidgley replied:
Agree! Will update.

domw30 commented on src/features/ui/Wrapper/Wrapper.module.scss:
```diff
@@ -0,0 +1,10 @@
+		justify-content: space-between;
+
+		.Label {
+			margin-left: 1.375rem;
```
Could use gap over margins here I think.

ashmidgley replied:
Doesn't seem to apply the left spacing if I use `gap` on `LabelContainer` here. Any suggestions?

domw30 commented on src/features/ui/Wrapper/Wrapper.tsx:
```diff
@@ -0,0 +1,30 @@
+	info: string;
+};
+
+export const Wrapper: FC<WrapperProps> = ({
```
At first look, I'm not 100% sure what `Wrapper` is used for. Would there be any alternatives to naming here?

ashmidgley replied:
Definitely agree! Have changed to `InputWrapper` to give a bit more context. Let me know if you think this fits the bill.

domw30 reviewed (changes requested):
Looking good! Some questions/comments. Could we also update `steps` folders to `Steps` :) 💯

domw30 commented:
@ashmidgley, would it be possible to resolve any comments/changes that have been fixed up just to reduce the noise above. Thanks mate!

ashmidgley commented:
Thanks for the review @domw30! Have pushed changes to address comments.

domw30 commented on src/features/create-dao/Steps/DetailsForm.test.tsx:
```diff
@@ -62,9 +62,7 @@ describe('<DetailsForm />', () => {
-			expect(
-				screen.getByText('The name field is required.'),
-			).not.toBeInTheDocument(),
+			expect(screen.getByText('The name field is required.')).toBeVisible(),
```
I'm not sure if the test description aligns with the expectation. Should the input validation error be visible if the required name field value is valid?

Could be wrong but looks like the test should be described as something like `should throw invalid input error`. I might just be reading this wrong!

Same for the tests below mate

domw30 commented on src/features/create-dao/Steps/GovernanceForm.test.tsx:
```diff
@@ -64,7 +64,7 @@ describe('<GovernanceForm />', () => {
 			expect(
 				screen.getByText('The voting period field is required.'),
-			).not.toBeInTheDocument(),
+			).toBeVisible(),
```
Please read comments above re testing using `isVisible`. Let me know if I've made sense/missed something

domw30 reviewed (approved):
Nice work Ash. Thanks for making those changes. Couple of questions to comment back on, but I'll approve. Let me know.  🚀

colbr commented:
Merging as @domw30 has approved! @ashmidgley can you chat with @domw30 about the above questions? Cheers 🎉
